File: taikyoku_engine/tsfen.py
"""TSFEN notation for Taikyoku Shogi positions.

Format: rank/rank/.../rank side move_number

Each rank lists pieces left to right. Empty squares use numbers (like FEN).
Piece format: Color prefix (uppercase=Black, lowercase=White) + abbreviation.
Multi-character abbreviations are wrapped: e.g., (CP) for Crown Prince.

Example fragment: L(TS)(RR)W... for Black's back rank.
"""

from .pieces import BOARD_SIZE, BLACK, WHITE, PIECE_NAME
from .board import TaikyokuBoard


def to_tsfen(board: TaikyokuBoard) -> str:
    """Convert board position to TSFEN string."""
    ranks = []
    for r in range(BOARD_SIZE):
        rank_str = _encode_rank(board, r)
        ranks.append(rank_str)

    side = 'b' if board.side_to_move == BLACK else 'w'
    return f"{'/'.join(ranks)} {side} {board.move_number}"


def from_tsfen(tsfen: str) -> TaikyokuBoard:
    """Parse TSFEN string into a TaikyokuBoard."""
    parts = tsfen.strip().split()
    rank_str = parts[0]
    side = parts[1] if len(parts) > 1 else 'b'
    move_num = int(parts[2]) if len(parts) > 2 else 1

    board = TaikyokuBoard()
    board.side_to_move = BLACK if side == 'b' else WHITE
    board.move_number = move_num

    ranks = rank_str.split('/')
    if len(ranks) != BOARD_SIZE:
        raise ValueError(f"Expected {BOARD_SIZE} ranks, got {len(ranks)}")

    for r, rank in enumerate(ranks):
        _decode_rank(board, r, rank)

    board._rebuild_indices()
    return board


def _encode_rank(board, row):
    """Encode a single rank as TSFEN."""
    parts = []
    empty_count = 0

    for c in range(BOARD_SIZE):
        cell = board.at(row, c)
        if cell is None:
            empty_count += 1
        else:
            if empty_count > 0:
                parts.append(str(empty_count))
                empty_count = 0
            piece, color = cell
            parts.append(_encode_piece(piece, color))

    if empty_count > 0:
        parts.append(str(empty_count))

    return ''.join(parts)


def _encode_piece(piece, color):
    """Encode a piece as TSFEN token.

    Black pieces: uppercase, White pieces: lowercase.
    Single-char abbreviations: just the char.
    Multi-char abbreviations: wrapped in parentheses.
    """
    if color == BLACK:
        s = piece.upper()
    else:
        s = piece.lower()

    if len(piece) == 1:
        return s
    else:
        return f"({s})"


def _decode_rank(board, row, rank_str):
    """Decode a TSFEN rank string and place pieces on the board."""
    col = 0
    i = 0
    n = len(rank_str)

    while i < n and col < BOARD_SIZE:
        ch = rank_str[i]

        if ch.isdigit():
            # Empty squares
            num_str = ''
            while i < n and rank_str[i].isdigit():
                num_str += rank_str[i]
                i += 1
            col += int(num_str)

        elif ch == '(':
            # Multi-character piece abbreviation
            i += 1  # skip '('
            abbrev = ''
            while i < n and rank_str[i] != ')':
                abbrev += rank_str[i]
                i += 1
            i += 1  # skip ')'

            # Determine color from case
            if abbrev[0].isupper() or (abbrev.startswith('+') and len(abbrev) > 1 and abbrev[1].isupper()):
                color = BLACK
                piece = abbrev.upper()
            else:
                color = WHITE
                piece = abbrev.upper()

            board.set_piece(row, col, piece, color)
            col += 1

        else:
            # Single character piece
            if ch.isupper():
                color = BLACK
                piece = ch
            else:
                color = WHITE
                piece = ch.upper()

            board.set_piece(row, col, piece, color)
            col += 1
            i += 1

File: src/tsfen.rs
//! TSFEN notation for Taikyoku Shogi positions.
//!
//! Format: `rank/rank/.../rank side move_number`
//!
//! - 36 ranks separated by `/`, from row 0 (top) to row 35 (bottom)
//! - Each rank lists pieces left to right; empty squares use numbers (like FEN)
//! - Black pieces: uppercase; White pieces: lowercase
//! - Single-char abbreviations: bare letter (e.g., `P`, `K`, `r`, `b`)
//! - Multi-char abbreviations: wrapped in parentheses (e.g., `(CP)`, `(dh)`)
//! - Side to move: `b` (Black) or `w` (White)
//! - Move number: integer starting at 1

use crate::types::*;
use crate::pieces;
use crate::board::Board;

/// Encode a board position as a TSFEN string.
pub fn to_tsfen(board: &Board) -> String {
    let mut ranks = Vec::with_capacity(BOARD_SIZE);
    for r in 0..BOARD_SIZE {
        ranks.push(encode_rank(board, r));
    }
    let side = if board.side_to_move == BLACK { 'b' } else { 'w' };
    format!("{} {} {}", ranks.join("/"), side, board.move_number)
}

/// Parse a TSFEN string into a Board. Returns `Err` with a message on failure.
pub fn from_tsfen(tsfen: &str) -> Result<Board, String> {
    let parts: Vec<&str> = tsfen.trim().splitn(3, ' ').collect();
    let rank_str = parts.first().ok_or("empty TSFEN")?;
    let side = parts.get(1).copied().unwrap_or("b");
    let move_num: u32 = parts.get(2)
        .and_then(|s| s.parse().ok())
        .unwrap_or(1);

    let mut board = Board::new();
    board.side_to_move = if side == "w" { WHITE } else { BLACK };
    board.move_number = move_num;

    let ranks: Vec<&str> = rank_str.split('/').collect();
    if ranks.len() != BOARD_SIZE {
        return Err(format!("expected {} ranks, got {}", BOARD_SIZE, ranks.len()));
    }

    for (r, rank) in ranks.iter().enumerate() {
        decode_rank(&mut board, r, rank)?;
    }

    board.rebuild_lists_pub();
    Ok(board)
}

fn encode_rank(board: &Board, row: usize) -> String {
    let mut parts = String::new();
    let mut empty = 0u32;

    for c in 0..BOARD_SIZE {
        let cell = board.cells[row * BOARD_SIZE + c];
        if cell == EMPTY_CELL {
            empty += 1;
        } else {
            if empty > 0 {
                parts.push_str(&empty.to_string());
                empty = 0;
            }
            let pt = cell_piece(cell);
            let color = cell_color(cell);
            let abbrev = pieces::abbrev(pt);
            encode_piece(&mut parts, abbrev, color);
        }
    }
    if empty > 0 {
        parts.push_str(&empty.to_string());
    }
    parts
}

fn encode_piece(out: &mut String, abbrev: &str, color: u8) {
    // Black = uppercase, White = lowercase
    let s: String = if color == BLACK {
        abbrev.to_uppercase()
    } else {
        abbrev.to_lowercase()
    };

    if abbrev.len() == 1 {
        out.push_str(&s);
    } else {
        out.push('(');
        out.push_str(&s);
        out.push(')');
    }
}

fn decode_rank(board: &mut Board, row: usize, rank_str: &str) -> Result<(), String> {
    let bytes = rank_str.as_bytes();
    let n = bytes.len();
    let mut col = 0usize;
    let mut i = 0usize;

    while i < n && col < BOARD_SIZE {
        let ch = bytes[i] as char;

        if ch.is_ascii_digit() {
            // Empty squares
            let start = i;
            while i < n && (bytes[i] as char).is_ascii_digit() {
                i += 1;
            }
            let count: usize = rank_str[start..i].parse()
                .map_err(|_| format!("invalid number in rank {}", row))?;
            col += count;
        } else if ch == '(' {
            // Multi-character abbreviation
            i += 1; // skip '('
            let start = i;
            while i < n && bytes[i] != b')' {
                i += 1;
            }
            if i >= n {
                return Err(format!("unclosed parenthesis in rank {}", row));
            }
            let abbrev_raw = &rank_str[start..i];
            i += 1; // skip ')'

            let (color, abbrev) = parse_piece_token(abbrev_raw);
            if let Some(pt) = pieces::find_by_abbrev(&abbrev) {
                let sq = row * BOARD_SIZE + col;
                board.cells[sq] = make_cell(pt, color);
            }
            col += 1;
        } else {
            // Single character piece
            let color = if ch.is_ascii_uppercase() { BLACK } else { WHITE };
            let abbrev = ch.to_ascii_uppercase().to_string();
            if let Some(pt) = pieces::find_by_abbrev(&abbrev) {
                let sq = row * BOARD_SIZE + col;
                board.cells[sq] = make_cell(pt, color);
            }
            col += 1;
            i += 1;
        }
    }
    Ok(())
}

/// Parse a piece token (from inside parens or single char) into (color, uppercase_abbrev).
fn parse_piece_token(token: &str) -> (u8, String) {
    let first_alpha = token.chars().find(|c| c.is_ascii_alphabetic());
    let color = match first_alpha {
        Some(c) if c.is_ascii_uppercase() => BLACK,
        _ => WHITE,
    };
    (color, token.to_ascii_uppercase())
}
